pyqtorch/analog.py

```diff
@@ -355,11 +335,17 @@ def __init__(
             operator = operator_to_sparse_diagonal(hamiltonian)
         else:
             operator = torch.diag(hamiltonian).reshape(-1, 1)
-        super().__init__(operator, operations.qubit_support[0])
-        self.qubit_support = operations.qubit_support
-        self.n_qubits = n_qubits
 
-    def run(self, state: Tensor, values: dict[str, Tensor]) -> Tensor:
+        # to make an observable with only one operation
+        # we represent this by a single Primitive
+        # with operator as the input tensor
+        # to be accessed later via self.operations[0].pauli
+        # and this is compatible with the to method for conversions
+        super().__init__(n_qubits, Primitive(operator, operations.qubit_support))
```

Comment:
This is ok for now to replicate the previous logic but this class will need to be reworked, essentially so we can deal with https://github.com/pasqal-io/pyqtorch/issues/175

I also don't think we need a special `DiagonalObservable` class, just a property that keeps track of the diagonalily of the individual blocks when they are composed of `Z`, `I` or `N` and a version of `apply_operator` specific for diagonal operations. This can be used throughout PyQ for applying diagonal evolutions, and also for applying diagonal gates like `CPHASE`.

But I would not do this in this MR.

Reply:
Sure, created issue #211 for handling this.